### PR TITLE
Updated the ItServerStartPolicy Tests to use new domain LifeCycle script instead of using patch commands

### PR DESCRIPTION
**Pull request by hzhao-github**
Update the test methods in ItServerStartPolicy class to use new domain LifeCycle script instead of using patch commands. See the coverage in ItSamples.java. The  ItSamples class does not cover independent managed server and configured cluster and AdminServer shutdown.

 Add some Negative Tests:
(a) can not shutdown admin server
(b) server/domain/cluster does not exists as passed thru command line
in a command line pass -c foo. ( where there is no such cluster foo in domain resource )
(c) start a manged server beyond limit
max Dynamic Cluster size is 3, you start a managed managed-server4
(d) shutdown Admin server and start/stop a manged server

padded Jenkins job:
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/3191 (parallel)
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/3193 (parallel)
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/3192 (sequential)

https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/3214 (parallel)
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/3219 (parallel)
https://build.weblogick8s.org:8443/job/weblogic-kubernetes-operator-kind-new/3218 (sequential)

### CODE REVIEW

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 1]
```diff
@@ -492,15 +548,18 @@ public void testConfigClusterStartServerUsingAlways() {
    * by patching the resource definition with 
    *  spec/managedServers/2/serverStartPolicy set to ALWAYS.
    * Make sure that managed server managed-server2 is up and running
+   * Verify that if server start policy is ALWAYS and the server is selected
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:551]: See the comment for testConfigClusterStartServerUsingAlways(), the same is applicable to this method

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 2]
```diff
@@ -624,36 +687,45 @@ public void testDynamicClusterReplicaCountIsMaintained() {
    * Make sure that ONLY the specified managed server is stopped. 
    * Restart the independent managed server by patching the resource definition 
    * with spec/managedServers/0/serverStartPolicy set to ALWAYS.
-   * Make sure that the specified managed server is in RUNNING state.
+   * Make sure that the specified managed server is in RUNNING state
+   * Verify that if server start policy is ALWAYS and the server is selected
+   * to start based on the replica count, it means that server is already started or is
+   * in the process of starting. In this case, script exits without making any changes.
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:693]: Remove the line 696 :  https://bug.oraclecorp.com/pls/bug/webbug_print.show?c_rptno=31833260. The BUG has been resolved.

Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:693]: Do not use patch method (patchServerStartPolicy)  to stop/start independent managed server.  Use the script to stop/start managed server.

Modify the method name from testStandaloneManagedRestartAlways() to testStandaloneManagedRestart()

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 3]
```diff
@@ -663,36 +735,175 @@ public void testStandaloneManagedRestartAlways() {
    * Make sure that ONLY the specified managed server is stopped. 
    * Restart the independent managed server by patching the resource definition 
    * with spec/managedServers/0/serverStartPolicy set to IF_NEEDED.
-   * Make sure that the specified managed server is in RUNNING state.
+   * Make sure that the specified managed server is in RUNNING state
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:738]: I think, the test testStandaloneManagedRestartAlways() covers this usecase. We can delete this method.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 4]
```diff
@@ -663,36 +735,175 @@ public void testStandaloneManagedRestartAlways() {
...
+
+    // Verify that if server start policy is IF_NEEDED and the server is selected
+    // to start based on the replica count, it means that server is already started or is
+    // in the process of starting. In this case, script exits without making any changes.
+    String result = executeLifecycleScript(START_SERVER_SCRIPT, SERVER_LIFECYCLE, serverName);
+    assertTrue(result.contains("No changes needed"), "startServer.sh shouldn't make changes");
+  }
+
+  /**
+   * Negative test to verify:
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:780]: non-exist --> non-existing

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 5]
```diff
@@ -663,36 +735,175 @@ public void testStandaloneManagedRestartAlways() {
...
+   * (a) the sample script can not stop or start a non-exist server
+   * (b) the sample script can not stop or start a non-exist cluster
+   * (c) the sample script can not stop or start a non-exist domain.
+   */
+  @Test
+  @DisplayName("Restart the configured cluster with serverStartPolicy")
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:786]: Modify the description for DisplayName()

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 6]
```diff
@@ -663,36 +735,175 @@ public void testStandaloneManagedRestartAlways() {
...
+        String.format("Failed to run %s", START_DOMAIN_SCRIPT));
+    assertTrue(verifyExecuteResult(result, regex),"The script shouldn't start a domain that doesn't exist");
+  }
+
+  /**
+   * Negative test to verify that the sample script can not start a server that exceeds the max cluster size
+   * e.g. the max cluster size is 2, the sample script can't start server 3
+   */
+  @Test
+  @DisplayName("Restart the configured cluster with serverStartPolicy")
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:841]: Modify the description for DisplayName()
Modify method name to testStartMSBeyondLimitNeg to testStartMangedServerBeyondMaxClusyterLimit()

Add a method for DynamicCluster size limit.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 7]
```diff
@@ -663,36 +735,175 @@ public void testStandaloneManagedRestartAlways() {
...
+
+    // verify that the script can not start a server that exceeds the max cluster size
+    String result =  assertDoesNotThrow(() ->
+        executeLifecycleScript(START_SERVER_SCRIPT, SERVER_LIFECYCLE, serverName, "", false),
+        String.format("Failed to run %s", START_SERVER_SCRIPT));
+    assertTrue(verifyExecuteResult(result, regex),"The script shouldn't stop a server that is beyong the limit");
+  }
+
+  /**
+   * Negative test to verify that after the admin server is stopped, the sample script can start or stop a server.
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:857]: Add the BUG number to the description

Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:857]: Change the description 
  In the absence of Administration Server, sample script can start/stop a managed server.

**Review by anpanigr — COMMENTED** [review]
Please take a look the the review comments

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 7 above)
Comment (vanajamukkara) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:857]: why is this a negative test? if you can start/stop a manager server without admin server, then it is a positive test.

**Review by vanajamukkara — CHANGES_REQUESTED** [review]

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 8]
```diff
@@ -454,29 +500,39 @@ public void testDomainRestart() {
    * by patching the resource definition with 
    *  spec/managedServers/1/serverStartPolicy set to ALWAYS.
    * Make sure that managed server config-cluster-server2 is up and running
+   * Verify that if server start policy is ALWAYS and the server is selected
```
[the remark this answers is not in the document]
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:503]: I understand that. That's why I just inserted 

    String result = executeLifecycleScript(START_SERVER_SCRIPT, SERVER_LIFECYCLE, serverName);
    assertTrue(result.contains("No changes needed"), "startServer.sh shouldn't make changes");

to

   // Verify that if server start policy is ALWAYS and the server is selected
    // to start based on the replica count, it means that server is already started or is
    // in the process of starting. In this case, script exits without making any changes.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 1 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:551]: same reason as above

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 4 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:780]: fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 9]
```diff
@@ -252,7 +280,8 @@ public void beforeEach() {
    * Make sure that Only the Administration server is stopped. 
    * Restart the Administration server by patching the resource definition with 
    *  spec/adminServer/serverStartPolicy set to IF_NEEDED.
-   * Make sure that the Administration server is in RUNNING state. 
+   * Make sure that the Administration server is in RUNNING state.
+   * Verify that the sample script, stopServer.sh can not shutdown admin server
```
[the remark this answers is not in the document]
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:284]: fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 10]
```diff
@@ -400,8 +443,12 @@ public void testDomainRestart() {
...
     String standaloneServerPodName = domainUid + "-standalone-managed";
 
-    patchServerStartPolicy("/spec/serverStartPolicy", "NEVER");
-    logger.info("Domain is patched to stop entire WebLogic domain");
+    //copy the samples directory to a temporary location
+    setupSample();
+
+    // Verify all WebLogic server instance pods are shut down after stopDomain script execution
+    logger.info("Stop entire WebLogic domain using the script");
+    executeLifecycleScript(STOP_DOMAIN_SCRIPT, DOMAIN, null);
```
[the remark this answers is not in the document]
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:451]: fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 5 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:786]: fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 6 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:841]: fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 7 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:857]: fixed

Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:857]: fixed

Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:857]: yea. you are right, it's not a negative test. fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 11]
```diff
@@ -291,15 +320,23 @@ public void testAdminServerRestart() {
     checkPodReadyAndServiceExists(adminServerPodName, 
             domainUid, domainNamespace);
     logger.info("AdminServer restart success");
+
+    //copy the samples directory to a temporary location
+    setupSample();
```
[the remark this answers is not in the document]
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java]: fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 4 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:756]: There are few more occurrences of the word "non-exist server"  such as line # 773, 779, 788

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 2 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:675]: This line has not been deleted in your recent commit

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 12]
```diff
@@ -663,36 +714,172 @@ public void testStandaloneManagedRestartAlways() {
...
+
+  /**
+   * Negative test to verify:
+   * (a) the sample script can not stop or start a non-existing server
+   * (b) the sample script can not stop or start a non-existing cluster
+   * (c) the sample script can not stop or start a non-existing domain.
+   */
+  @Test
+  @DisplayName("Verify that the sample script can not stop or start non-existing components")
+  public void testRestartWrongComponent() {
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:763]: Change the name to testRestartNonExistingComponent()

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 13]
```diff
@@ -663,36 +714,172 @@ public void testStandaloneManagedRestartAlways() {
...
+    assertTrue(verifyExecuteResult(result, regex),"The script shouldn't start a domain that doesn't exist");
+  }
+
+  /**
+   * Negative test to verify that the sample script can not start a server that exceeds the max cluster size
+   * e.g. the max cluster size is 2, the sample script can't start server 3
+   */
+  @Test
+  @DisplayName("Verify that the sample script can not start a server that exceeds the max cluster size")
+  public void testStartMangedServerBeyondMaxClusyterLimit() {
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:816]: typo:  testStartMangedServerBeyondMaxClusyterLimit() --> testStartMangedServerBeyondMaxClusterLimit()

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 14]
```diff
@@ -663,36 +714,172 @@ public void testStandaloneManagedRestartAlways() {
...
+    assertTrue(verifyExecuteResult(result, regex),"The script shouldn't stop a server that is beyong the limit");
+  }
+
+  /**
+   * Verify that after the admin server is stopped, the sample script can start or stop a server.
+   */
+  @Disabled("Due to the bug OWLS-86251")
+  @Test
+  @DisplayName("In the absence of Administration Server, sample script can start/stop a managed server")
+  public void testServerRestartMSWithoutAdmin() {
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:841]: testServerRestartMSWithoutAdmin --> testServerRestartManagedServerWithoutAdmin

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 15]
```diff
@@ -663,36 +714,172 @@ public void testStandaloneManagedRestartAlways() {
...
+    assertTrue(verifyExecuteResult(result, regex),"The script shouldn't stop a domain that doesn't exist");
+
+    // verify that the script can not start a non-exist domain
+    result = assertDoesNotThrow(() ->
+        executeLifecycleScript(START_DOMAIN_SCRIPT, DOMAIN, null, "", false, domainName),
+        String.format("Failed to run %s", START_DOMAIN_SCRIPT));
+    assertTrue(verifyExecuteResult(result, regex),"The script shouldn't start a domain that doesn't exist");
+  }
+
+  /**
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:810]: Describe the configuration details with usecase as follows ...
    Currently, the domain resource has a configured cluster with two managed servers and a dynamic cluster with MaxClusterSize set to 5.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 14 above)
Comment (anpanigr) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:841]: Add a test to start/stop a manged server form a dynamic cluster.

**Review by anpanigr — CHANGES_REQUESTED** [review]
Please address some unresolved comments.

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 4 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:756]: fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 13 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:816]: fixed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 12 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:763]: changed

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 15 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:810]: added

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 16]
```diff
@@ -752,6 +746,27 @@ public void testStandaloneManagedRestartIfNeeded() {
...
+   * Start a dynamic cluster managed server within the max cluster size
+   * and verify the replica count increased.
+   */
+  @Test
+  @DisplayName("Start a dynamic cluster managed server within the max cluster size and verify replica number increased")
+  public void testDynamicClusterReplicaCountIncrease() {
+    String serverName = "managed-server3";
+    String serverPodName = domainUid + "-" + serverName;
+    String serverPodName2 = domainUid + "-managed-server2";
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:758]: unused variable serverPodName2

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java [hunk 17]
```diff
@@ -752,6 +746,27 @@ public void testStandaloneManagedRestartIfNeeded() {
     assertTrue(result.contains("No changes needed"), "startServer.sh shouldn't make changes");
   }
 
+  /**
+   * Start a dynamic cluster managed server within the max cluster size
+   * and verify the replica count increased.
```
Comment (anpanigr) [inline integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:751]: Here we do not check the repilca count. Here we can check the stdout of the startupScript which indicate that replica count has been increased OR we can create a method to get the updated replica count from the domain resource.

**Review by anpanigr — CHANGES_REQUESTED** [review]
Please update the class level javadoc to reflect the new/modified usecases

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 14 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java]: done

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 16 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:758]: it's used at the line 761

File: integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java (hunk 17 above)
Comment (hzhao-github) [reply integration-tests/src/test/java/oracle/weblogic/kubernetes/ItServerStartPolicy.java:751]: changed

**Review by anpanigr — APPROVED** [review]
LGTM. update the PR description with the latest jenkin result.

**Review by vanajamukkara — APPROVED** [review]

**Pull request merged**